decide-app/src/main/java/com/decide/app/LaunchInterceptorConditions.java

```diff
@@ -301,6 +301,37 @@ public double distance(Point p1, Point p2) {
         return Math.sqrt(dx*dx + dy*dy);
     }
 
+    /**
+     * Tell whether or not three points can be contained on or inside a
+     * circle with specified radius.
+     * @param p1 The first point.
+     * @param p2 The second point.
+     * @param p3 The third point.
+     * @param r The radius of the circle.
+     * @return True if all points can be contained on or inside the circle.
+     */
+    public boolean containedInCircle(Point p1, Point p2, Point p3, double R) {
+        if (p1 == null || p2 == null || p3 == null) throw new IllegalArgumentException();
+        if (Double.isNaN(R) || R < 0) throw new IllegalArgumentException();
+        double d1 = distance(p1, p2);
+        double d2 = distance(p1, p3);
+        double d3 = distance(p2, p3);
+        if (d1 > 2*R || d2 > 2*R || d3 > 2*R) return false;
+
+        // Midpoints (Act as centers)
+        Point m12 = new Point(p1.getX()/2 + p2.getX()/2, p1.getY()/2 + p2.getY());
+        Point m23 = new Point(p2.getX()/2 + p3.getX()/2, p2.getY()/2 + p3.getY());
+        Point m13 = new Point(p1.getX()/2 + p3.getX()/2, p1.getY()/2 + p3.getY());
+        if (distance(m12, p3) <= R) return true;

```

Comment:
Are you sure of this part? Do you have a source?
Otherwise I think it can be done by finding the point at the intersection of the lines passing through the middle points and perpendicular to their respecting lines (the circumcenter, as explained here https://www.storyofmathematics.com/triangle-inside-a-circle/)

Reply:
It might not be a correct approach if as in the tests you accept points on a line. Yours would thus be better.

Reply:
> *A triangle inside a circle is a triangle where all three vertices lie on the circumference of the circle.*

The problem is that not all points have to lie on the circumference.

The method is divided into three parts:
1. If two points are farther than 2*R apart, then it cannot be contained with just radius R.
2. There is a possibility that only 2 points are on the circumference. Since all pair-wise distances are less than or equal to 2*R, we now assume that two points will be on the circumference of the circle (opposite to each other). The center is the midpoint on the line straight from one point to the other. If we distance from the center to the third point is less than or equal to R, then it can be contained. We do this check rotating between which 2 points are on the circumference.
3. Lastly, since the previous check failed, all three points must be on the circumference, which is calculated similarly to what you linked to.

The midpoints were not part of the original implementation in LIC8, but as I discovered when testing the method separately, it would then be unable to find that this is containable by the blue circle, since it would find that the smallest circle to contain them with is the red circle, which is larger than it has to be:

![bild](https://github.com/alexarne/DECIDE/assets/75060613/4bfef6c7-4d1c-4196-8555-df5a1714201f)

Reply:
As for *how* we calculate the midpoints, it is simply derived from:
`v_1 + (v_2-v_1)/2 = 2*v_1/2 + v_2/2 - v_1/2 = v_1/2 + v_2/2`